catalog/core/catalog-core-versioning/versioning-common/src/main/java/ddf/catalog/core/versioning/MetacardVersion.java

```diff
@@ -209,31 +309,82 @@ public MetacardVersion(Metacard sourceMetacard, Action action, Subject subject)
         this.setTags(Collections.singleton(VERSION_TAG));
     }
 
+    private byte[] getVersionType(Metacard sourceMetacard) {
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+            oos.writeObject(sourceMetacard.getMetacardType());
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException("Could not serialize MetacardType", e);
+        }
+    }
+
     /**
-     * Returns a {@link BasicTypes#BASIC_METACARD} version of the given
-     * {@link MetacardVersion}
+     * Reverts the {@link MetacardVersion} to the original {@link Metacard}
      *
      * @return The converted metacard
      * @throws IllegalStateException
      */
-    public Metacard toBasicMetacard() {
-        return toBasicMetacard(this);
+    public Metacard getMetacard(List<MetacardType> types) {
+        return toMetacard(this, types);
     }
 
-    public static Metacard toBasicMetacard(Metacard source) {
+    public static Metacard toMetacard(Metacard source, List<MetacardType> types) {
         String id = (String) source.getAttribute(MetacardVersion.VERSION_OF_ID)
                 .getValue();
         if (isNullOrEmpty(id)) {
             throw new IllegalStateException(
                     "Cannot convert history metacard without the original metacard id");
         }
 
-        MetacardImpl result = new MetacardImpl(source, BasicTypes.BASIC_METACARD);
+        String typeString = (String) source.getAttribute(VERSION_TYPE)
+                .getValue();
+        Optional<MetacardType> typeFromExisting = types.stream()
+                .filter(mt -> mt.getName()
+                        .equals(typeString))
+                .findFirst();
+
+        MetacardImpl result = new MetacardImpl(source,
+                typeFromExisting.orElseGet(() -> getMetacardTypeBinary(source).orElseThrow(
```

Comment:
I don't like this inlined retrieval of type/fallback/throw. I'd rather you pulled this out to the stream above and change the `Optional<MetacardType>` to a `MetacardType`. I believe it will read more cleanly that way.